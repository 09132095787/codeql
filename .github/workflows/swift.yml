name: "Swift"

on:
  pull_request:
    paths:
      - "swift/**"
      - "misc/bazel/**"
      - "*.bazel*"
      - .github/workflows/swift.yml
      - .github/actions/fetch-codeql/action.yml
      - codeql-workspace.yml
      - .pre-commit-config.yaml
      - "!**/*.md"
      - "!**/*.qhelp"
    branches:
      - main
defaults:
  run:
    working-directory: swift

jobs:
  changes:
    runs-on: ubuntu-latest
    outputs:
      codegen: ${{ steps.filter.outputs.codegen }}
      ql: ${{ steps.filter.outputs.ql }}
    steps:
      - uses: dorny/paths-filter@v2
        id: filter
        with:
          filters: |
            codegen:
              - 'github/workflows/swift.yml'
              - "misc/bazel/**"
              - "*.bazel*"
              - 'swift/actions/env-setup/**'
              - '.pre-commit-config.yaml'
              - 'swift/codegen/**'
              - 'swift/schema.py'
              - 'swift/**/*.dbscheme'
              - 'swift/ql/lib/codeql/swift/elements.qll'
              - 'swift/ql/lib/codeql/swift/elements/**'
              - 'swift/ql/lib/codeql/swift/generated/**'
              - 'swift/ql/test/extractor-tests/generated/**'
            ql:
              - 'github/workflows/swift.yml'
              - 'swift/**/*.ql'
              - 'swift/**/*.qll'
  # not using a matrix as you cannot depend on a specific job in a matrix, and we want to start linux checks
  # without waiting for the macOS build
  build-macos:
    runs-on: macos-12-xl
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/create-extractor-pack
      # loading bazel targets is very long on the basic macOS runner
      # therefore it's better to test the X
      - name: Test the Xcode autobuilder
        run: |
          bazel test //swift/xcode-autobuilder/tests
  build-linux:
    runs-on: ubuntu-20.04
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/create-extractor-pack
  qltests-linux:
    needs: build-linux
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/run-ql-tests
  qltests-macos:
    needs: build-macos
    runs-on: macos-12-xl
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/run-ql-tests
  integration-tests-linux:
    needs: build-linux
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/run-integration-tests
  integration-tests-macos:
    needs: build-macos
    runs-on: macos-12-xl
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/run-integration-tests
  codegen:
    runs-on: ubuntu-latest
    needs: changes
    if: ${{ needs.changes.outputs.codegen == 'true' }}
    steps:
      - uses: actions/checkout@v3
      - uses: ./swift/actions/env-setup
      - uses: pre-commit/action@v3.0.0
        name: Check that python code is properly formatted
        with:
          extra_args: autopep8 --all-files
      - name: Run unit tests
        run: |
          bazel test //swift/codegen/test --test_output=errors
      - uses: ./.github/actions/fetch-codeql
      - uses: pre-commit/action@v3.0.0
        name: Check that QL generated code was checked in
        with:
          extra_args: swift-codegen --all-files
      - name: Generate C++ files
        run: |
          bazel run //swift/codegen:codegen -- --generate=trap,cpp --cpp-output=$PWD/generated-cpp-files
      - uses: actions/upload-artifact@v3
        with:
          name: swift-generated-cpp-files
          path: swift/generated-cpp-files/**
  qlformat:
    runs-on: ubuntu-latest
    needs: changes
    if: ${{ needs.changes.outputs.ql == 'true' }}
    steps:
      - uses: actions/checkout@v3
      - uses: ./.github/actions/fetch-codeql
      - name: Check QL formatting
        run: find ql "(" -name "*.ql" -or -name "*.qll" ")" -print0 | xargs -0 codeql query format --check-only
