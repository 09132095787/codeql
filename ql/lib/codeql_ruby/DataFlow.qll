/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) data flow analyses.
 */
module DataFlow {
  import codeql_ruby.dataflow.internal.DataFlowImpl
}
