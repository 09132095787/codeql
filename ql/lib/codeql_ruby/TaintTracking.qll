/**
 * Provides classes for performing local (intra-procedural) and
 * global (inter-procedural) taint-tracking analyses.
 */
module TaintTracking {
  import codeql_ruby.dataflow.internal.tainttracking1.TaintTrackingImpl
}
