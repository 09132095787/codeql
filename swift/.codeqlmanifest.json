{
  "provide": [
    "ql/lib/qlpack.yml",
    "ql/test/qlpack.yml",
    "extractor_pack/codeql-extractor.yml"
  ]
}
