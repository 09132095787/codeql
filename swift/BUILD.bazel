load("@rules_pkg//:mappings.bzl", "pkg_attributes", "pkg_filegroup", "pkg_files")
load("@ql//:defs.bzl", "codeql_platform")

pkg_files(
    name = "dbscheme",
    srcs = [
        "ql/lib/swift.dbscheme",
        "ql/lib/swift.dbscheme.stats",
    ],
)

pkg_files(
    name = "qltest",
    srcs = ["tools/qltest.sh"],
    attributes = pkg_attributes(mode = "0755"),
    prefix = "tools",
)

pkg_files(
    name = "manifest",
    srcs = ["codeql-extractor.yml"],
)

pkg_filegroup(
    name = "extractor-pack-generic",
    srcs = [
        ":dbscheme",
        ":manifest",
        ":qltest",
    ],
    visibility = ["//visibility:public"],
)

pkg_files(
    name = "extractor",
    srcs = ["//swift/extractor"],
    attributes = pkg_attributes(mode = "0755"),
    prefix = "tools/" + codeql_platform,
)

pkg_filegroup(
    name = "extractor-pack-arch",
    srcs = [":extractor"],
    visibility = ["//visibility:public"],
)

pkg_filegroup(
    name = "extractor-pack",
    srcs = [
        ":extractor-pack-arch",
        ":extractor-pack-generic",
    ],
    visibility = ["//visibility:public"],
)
