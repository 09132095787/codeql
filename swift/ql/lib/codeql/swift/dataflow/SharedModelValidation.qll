/**
 * INTERNAL: Do not use.
 *
 * Provides classes for validating kinds in models as data rows.
 * Such that we can share this logic across our CodeQL analysis of different languages.
 */
class ValidSinkKind extends string {
  ValidSinkKind() {
    this =
      [
        // shared ALL languages
        "request-forgery", "ldap-injection", "sql-injection", "nosql-injection", "log-injection",
        "xpath-injection", "html-injection", "js-injection", "url-redirection", "path-injection",
        "file-content-store", "hostname-verification", "response-splitting", "information-leak",
        "xslt-injection", "template-injection", "fragment-injection", "command-injection",
        "unsafe-deserialization", "xxe", "database-store", "format-string",
        // .matches("credentials-%"), .matches("regex-use%")"
        // shared MOST languages
        "code-injection", // .matches("encryption-%"),
        // Java only
        "jndi-injection", "mvel-injection", "groovy-injection", "ognl-injection", "jexl-injection",
        "bean-validation", "intent-redirection", "pending-intents",
        // JS only
        "mongodb.sink",
        // Swift only
        "preferences-store", "transmission", "predicate-injection", "webview-fetch",
        "tls-protocol-version", "hash-iteration-count" // .matches("%string-%length"), .matches("weak-hash-input-")
      ]
  }
}
