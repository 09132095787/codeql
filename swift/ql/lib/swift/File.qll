/** TODO: to be replaced */

/** A file. */
class File extends @file {
  /** toString */
  string toString() { result = getAbsolutePath() }

  /** Gets the name of this file. */
  string getName() { files(this, result) }

  /** Gets the absolute path of this file. */
  string getAbsolutePath() { result = getName() }

  /** Gets the full name of this file. */
  string getFullName() { result = getAbsolutePath() }

  /** Gets the URL of this file. */
  string getURL() { result = "file://" + this.getAbsolutePath() + ":0:0:0:0" }

  /** Gets the base name of this file. */
  string getBaseName() {
    result = this.getAbsolutePath().regexpCapture(".*/(([^/]*?)(?:\\.([^.]*))?)", 1)
  }
}
