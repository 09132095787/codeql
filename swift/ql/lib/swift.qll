/** Top-level import for the Swift language pack */

import swift.File
