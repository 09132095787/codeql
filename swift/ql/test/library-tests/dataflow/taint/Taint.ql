/**
 * @kind path-problem
 */

import swift
import codeql.swift.dataflow.TaintTracking
import codeql.swift.dataflow.DataFlow::DataFlow
import PathGraph

class TestConfiguration extends TaintTracking::Configuration {
  TestConfiguration() { this = "TestConfiguration" }

  override predicate isSource(Node src) {
    src.asExpr().(CallExpr).getStaticTarget().getName() = "source()"
  }

  override predicate isSink(Node sink) {
    exists(CallExpr sinkCall |
      sinkCall.getStaticTarget().getName() = "sink(arg:)" and
      sinkCall.getAnArgument().getExpr() = sink.asExpr()
    )
  }

  override int explorationLimit() { result = 100 }
}

from PathNode src, PathNode sink, TestConfiguration test
where test.hasFlowPath(src, sink)
select sink, src, sink, "result"
