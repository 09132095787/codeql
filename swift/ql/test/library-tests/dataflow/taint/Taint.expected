edges
| string.swift:5:11:5:18 | call to source() :  | string.swift:7:13:7:13 | "..." |
| string.swift:5:11:5:18 | call to source() :  | string.swift:9:13:9:13 | "..." |
| string.swift:5:11:5:18 | call to source() :  | string.swift:11:13:11:13 | "..." |
| string.swift:5:11:5:18 | call to source() :  | string.swift:16:13:16:13 | "..." |
| string.swift:5:11:5:18 | call to source() :  | string.swift:18:13:18:13 | "..." |
nodes
| string.swift:5:11:5:18 | call to source() :  | semmle.label | call to source() :  |
| string.swift:7:13:7:13 | "..." | semmle.label | "..." |
| string.swift:9:13:9:13 | "..." | semmle.label | "..." |
| string.swift:11:13:11:13 | "..." | semmle.label | "..." |
| string.swift:16:13:16:13 | "..." | semmle.label | "..." |
| string.swift:18:13:18:13 | "..." | semmle.label | "..." |
subpaths
#select
| string.swift:7:13:7:13 | "..." | string.swift:5:11:5:18 | call to source() :  | string.swift:7:13:7:13 | "..." | result |
| string.swift:9:13:9:13 | "..." | string.swift:5:11:5:18 | call to source() :  | string.swift:9:13:9:13 | "..." | result |
| string.swift:11:13:11:13 | "..." | string.swift:5:11:5:18 | call to source() :  | string.swift:11:13:11:13 | "..." | result |
| string.swift:16:13:16:13 | "..." | string.swift:5:11:5:18 | call to source() :  | string.swift:16:13:16:13 | "..." | result |
| string.swift:18:13:18:13 | "..." | string.swift:5:11:5:18 | call to source() :  | string.swift:18:13:18:13 | "..." | result |
